src/drivers/Watchdog.cpp

```diff
@@ -2,74 +2,148 @@
+    value = (reason >> 17u) & 0x01u;
+    if(value != 0) {
+      return Watchdog::ResetReasons::LpComp;
+    }
+
+    value = (reason >> 18u) & 0x01u;
+    if(value != 0) {
+      return Watchdog::ResetReasons::DebugInterface;
+    }
+
+    value = (reason >> 19u) & 0x01u;
+    if(value != 0) {
+      return Watchdog::ResetReasons::NFC;
+    }
+
+    return Watchdog::ResetReasons::HardReset;
+  }
 }
 
-void Watchdog::Start() {
-  NRF_WDT->TASKS_START = 1;
+void Watchdog::Setup(uint8_t timeoutSeconds, SleepBehaviours sleepBehaviour, HaltBehaviours haltBehaviour) {
+  SetBehaviours(sleepBehaviour, haltBehaviour);
+  SetTimeout(timeoutSeconds);
+  EnableFirstReloadRegister();
+
+  resetReason = GetResetReason();
 }
 
-void Watchdog::Kick() {
-  NRF_WDT->RR[0] = WDT_RR_RR_Reload;
+void Watchdog::Start() {
+  // Write 1 in the START task to start the watchdog
+  NRF_WDT->TASKS_START = 1;
 }
 
-Watchdog::ResetReasons Watchdog::ActualResetReason() const {
-  uint32_t reason = NRF_POWER->RESETREAS;
-  NRF_POWER->RESETREAS = 0xffffffff;
-
-  if (reason & 0x01u)
-    return ResetReasons::ResetPin;
-  if ((reason >> 1u) & 0x01u)
-    return ResetReasons::Watchdog;
-  if ((reason >> 2u) & 0x01u)
-    return ResetReasons::SoftReset;
-  if ((reason >> 3u) & 0x01u)
-    return ResetReasons::CpuLockup;
-  if ((reason >> 16u) & 0x01u)
-    return ResetReasons::SystemOff;
-  if ((reason >> 17u) & 0x01u)
-    return ResetReasons::LpComp;
-  if ((reason) &0x01u)
-    return ResetReasons::DebugInterface;
-  if ((reason >> 19u) & 0x01u)
-    return ResetReasons::NFC;
-  return ResetReasons::HardReset;
+void Watchdog::Reload() {
+  // Write the reload value 0x6E524635UL to the reload register to reload the watchdog.
+  // NOTE : This driver enables only the 1st reload register.
+  NRF_WDT->RR[0] = ReloadValue;
 }
 
-const char* Watchdog::ResetReasonToString(Watchdog::ResetReasons reason) {
+const char* Pinetime::Drivers::ResetReasonToString(Watchdog::ResetReasons reason) {
   switch (reason) {
-    case ResetReasons::ResetPin:
+    case Watchdog::ResetReasons::ResetPin:
       return "Reset pin";
-    case ResetReasons::Watchdog:
+    case Watchdog::ResetReasons::Watchdog:
       return "Watchdog";
-    case ResetReasons::DebugInterface:
+    case Watchdog::ResetReasons::DebugInterface:
       return "Debug interface";
-    case ResetReasons::LpComp:
+    case Watchdog::ResetReasons::LpComp:
       return "LPCOMP";
-    case ResetReasons::SystemOff:
+    case Watchdog::ResetReasons::SystemOff:
       return "System OFF";
-    case ResetReasons::CpuLockup:
+    case Watchdog::ResetReasons::CpuLockup:
       return "CPU Lock-up";
-    case ResetReasons::SoftReset:
+    case Watchdog::ResetReasons::SoftReset:
       return "Soft reset";
-    case ResetReasons::NFC:
+    case Watchdog::ResetReasons::NFC:
       return "NFC";
-    case ResetReasons::HardReset:
+    case Watchdog::ResetReasons::HardReset:
       return "Hard reset";
     default:
       return "Unknown";
```

Comment:
This isn't a very safe way to return a string because the string is stack-allocated. I would suggest an array of strings in the anonymous namespace, and returning the string at the index of the ResetReason.

Reply:
I'm pretty sure this just returns a pointer to static data.

Reply:
After looking it up, it does seem that string literals have static duration. I was taught that you shouldn't return a string literal, but I guess that isn't correct.